Src/main.c

```diff
@@ -100,14 +107,36 @@ int main(void)
   debug("\n\nBootcamp starting up...");
   debug("Compiled on %s at %s", __DATE__, __TIME__);
 
-
   HAL_GPIO_WritePin(LED1_GPIO_Port, LED1_Pin, GPIO_PIN_SET);
+
+  HAL_ADC_Start(&hadc); //Enables ADC and starts conversion of the regular channels.
+  HAL_TIM_PWM_Start(&htim16, TIM_CHANNEL_1); //starts PWM signal generation. First argument is the pointer to the TIM handle and the second parameter is the channel used or the timer
+
+
   /* USER CODE END 2 */
 
   /* Infinite loop */
   /* USER CODE BEGIN WHILE */
+
+  struct bit_field b; //defines bitfield which will contain the 12 bit adc_output
+  b.adc_output = 0; 
+
+  //int adc_output = 0; //ADC gives a 12 bit output 
+
+  uint16_t pwm_compare = 0; //we only need to provide a 16 bit input for the pwm
+
   while (1)
   {
+    b.adc_output = HAL_ADC_GetValue(&hadc); //Gets the converted value from data register of regular channel. (A value between 0 and 4095 representinng 0 to 3.3 V [assuming a 12-bit ADC is being used])
+    /*
+    
+      From my understanding, the __HAL_TIM_SET_COMPARE() function compares the pwm_compare value to the period of the TIM16. The value a = pwm_compare/TIM16.Period*100% is the duty period of the PWM.
+      Since we want 1ms to 2ms on-time, we are looking at a 5-10% duty cycle. Right now, the PWM period is 60000, thus we want the pwm_compare value to have a range of 3000 to 6000!
+
+    */
+    pwm_compare = b.adc_output * 3000 / 4095 + 3000; // Converts value to a range of 3000 (1 ms on-time) to 6000 (2 ms on-time)
```

Comment:
Few notes here: watch out for inadvertently truncating some results. I made this mistake (looking back, I didn't actually fix it fully haha). If you plug in an adc_output, say 5, into your equation, pwm_compare should = 3006, but you return 3005. If you try, say 11, pwm_compare == 3008, which is correct. This is due to 2 factors working at the same time:

1st:
- Think about the output of the statement that you have there. What happens if there are decimal places left in this answer?

2nd:
- This error was one that I also made, and it took me a hot second to figure it out. [This link explains your problem nicely](https://www.quora.com/In-C-integer-division-truncates-Please-explain-this-line#:~:text=When%20integers%20are%20divided%20in,is%20ceiling%20of%20the%20output%20)

I'll leave the solution for the second factor to you to figure out, but the [hint I'll give is to look into data types, and what happens to the decimal points if you divide them](https://docs.microsoft.com/en-us/cpp/c-language/usual-arithmetic-conversions?view=vs-2019) :)

Note: in both of these cases, the term that you can refer to for this problem is called **truncation** on your stack overflow adventures.

Reply:
Okay, so to overcome this (refer to most recent commit), I first calculated the PWMCompare value as a float (stored in PWMCompareDecimal) which I then round. I then cast the float to uint16_t when assigning its value to PWMCompare. 

I believe that should fix it.